File: java/app-encryption/src/main/java/com/godaddy/asherah/crypto/keys/SharedCryptoKey.java
package com.godaddy.asherah.crypto.keys;

import java.time.Instant;
import java.util.function.Function;
import java.util.function.Consumer;

class SharedCryptoKey extends CryptoKey {
  private final CryptoKey sharedKey;

  SharedCryptoKey(final CryptoKey sharedKey) {
    this.sharedKey = sharedKey;
  }

  CryptoKey getSharedKey() {
    return sharedKey;
  }

  @Override
  public Instant getCreated() {
    return sharedKey.getCreated();
  }

  @Override
  public boolean isRevoked() {
    return sharedKey.isRevoked();
  }

  @Override
  public void markRevoked() {
    sharedKey.markRevoked();
  }

  @Override
  public void withKey(final Consumer<byte[]> action) {
    sharedKey.withKey(action);
  }

  @Override
  public <T> T withKey(final Function<byte[], T> action) {
    return sharedKey.withKey(action);
  }

  @Override
  public void close() {
    //SharedCryptoKey doesn't *own* any secrets so it doesn't have anything to close
  }
}

File: csharp/AppEncryption/Crypto/Keys/SharedCryptoKey.cs
using System;
using System.Runtime.CompilerServices;

[assembly: InternalsVisibleTo("AppEncryption.Tests")]
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1816:Call GC.SuppressFinalize correctly", Justification = "This class does not have a finalizer and does not need to suppress finalization.")]

namespace GoDaddy.Asherah.Crypto.Keys
{
    public class SharedCryptoKey : CryptoKey
    {
        internal SharedCryptoKey(CryptoKey sharedKey)
        {
            SharedKey = sharedKey;
        }

        internal CryptoKey SharedKey { get; }

        /// <inheritdoc />
        public override DateTimeOffset GetCreated()
        {
            return SharedKey.GetCreated();
        }

        /// <inheritdoc />
        public override void WithKey(Action<byte[]> actionWithKey)
        {
            SharedKey.WithKey(actionWithKey);
        }

        /// <inheritdoc />
        public override TResult WithKey<TResult>(Func<byte[], TResult> actionWithKey)
        {
            return SharedKey.WithKey(actionWithKey);
        }

        /// <inheritdoc />
        public override void Dispose()
        {
            // SharedCryptoKey doesn't *own* any secrets so it doesn't have anything to dispose
        }

        /// <inheritdoc />
        public override bool IsRevoked()
        {
            return SharedKey.IsRevoked();
        }

        /// <inheritdoc />
        public override void MarkRevoked()
        {
            SharedKey.MarkRevoked();
        }
    }
}
